workshops/full-stack_todo_app/README.md

```diff
@@ -0,0 +1,749 @@
+})
+```
+
+⭐ **Important**: The `value` event is called every time data is changed at the specified database reference, including changes to children. To limit the size of your snapshots, attach only at the lowest level needed for watching changes. For example, attaching a listener to the root of your database is not recommended.
+
+Now we need the data as an array to put it into the state. Declare an empty array and add the `id` and `task` into the array.
+
+```jsx
+todoRef.on("value", (snapshot) => {
+  const todos = snapshot.val()
+  const todoList = []
+  for(let id in todos){
+    todoList.push({id,...todos[id]})            // <------------ adding todos to todoList
+  }
+  setTodoList(todoList)         // <----------- setting the state
+})
+```
+
+Now that we have the data in `todoList`, we can map through it to display it.
+
+Initially, it is an empty array. So we have to check whether it is empty or not, then map through it. 
+
+For now, let's display using `<h1>`. Later we create a component for saying it.
+
+```jsx
+return(
+  <div className="todo-list">
+    {todoList ? todoList.map((todo) => <h1> {todo.task}</h1>) : null}
+  </div>
+)
+```
+
+<details>
+<summary>Your TodoList component so far looks like:</summary>
+
+```jsx
+import React, { useState, useEffect } from 'react'
+import firebase from '../Firebase'
+
+function TodoList(){
+  const [todoList, setTodoList] = useState([])
+
+  useEffect(() => {
+    const todoRef = firebase.database().ref('todo')
+    // Syncronizing
+    todoRef.on("value", (snapshot) => {
+      const todos = snapshot.val()
+      const todoList = []
+      for(let id in todos){
+        todoList.push({id,...todos[id]})
+      }
+      setTodoList(todoList)
+    })
+
+  }, [])
+  
+  return(
+    <div className="todo-list">
+      {todoList ? todoList.map((todo) => <h1> {todo.task}</h1>) : null}
+    </div>
+  )
+}
+
+export default TodoList
+```
+
+</details>   
+
+
+This will display the data that is retrieved from the database. Make sure you import and render the `TodoList` component into `App.js`.
+
+![Retrived data](https://cloud-792qf1oid.vercel.app/2reading_database.png)
+
+So far, we have done **Create** and **Read** operations. Halfway done!!! 🙌
+
+![Half way done](https://cloud-792qf1oid.vercel.app/7half_way_done_.gif)
+
+Let's continue to work with **Update** and **Delete** operations.
+
+### Creating Todo Item
+
+Create another file, `Todo.js`, in the `components` folder. Import React and Firebase. Also, create a functional component that takes a prop `todo`.
+
+```jsx
+import React from 'react'.
+import firebase from '../Firebase'
+
+function Todo({ todo }){
+  return(
+    <div classsName="todo-item">
+
+    </div>
+  )
+}
+
+export default Todo
+```
+
+Update the `TodoList` by importing `Todo` and rendering it.
+
+```jsx
+return(
+  <div className="todo-list">
+    {todoList ? todoList.map((todo, i) => <Todo todo={ todo } key={ i } />) : null}
+  </div>
+)
```

Comment:
can you please explain what we are doing here. Because, i know that we are rendering a todo component for each todo and also passing the data to that component as prop, but beginners will find it confusing, please add a short explanation on this